File: apps/api/internal/adapter/driven/mongodb/notification_repository.go
package mongodb

import (
	"context"
	"time"

	"github.com/neurodynecorp/api/internal/domain/entity"
	"github.com/neurodynecorp/api/internal/domain/port"
	"github.com/neurodynecorp/api/pkg/metrics"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
)

type NotificationRepository struct {
	collection *mongo.Collection
	metrics    *metrics.Metrics
}

func NewNotificationRepository(client *Client, m *metrics.Metrics) *NotificationRepository {
	return &NotificationRepository{
		collection: client.Collection("notifications"),
		metrics:    m,
	}
}

func (r *NotificationRepository) Create(ctx context.Context, notif *entity.Notification) error {
	start := time.Now()
	defer func() {
		r.metrics.MongoLatency.WithLabelValues("notifications", "insert").Observe(time.Since(start).Seconds())
		r.metrics.MongoOperations.WithLabelValues("notifications", "insert").Inc()
	}()
	_, err := r.collection.InsertOne(ctx, notif)
	return err
}

func (r *NotificationRepository) GetByUser(ctx context.Context, userID string, page, pageSize int) ([]*entity.Notification, int64, error) {
	start := time.Now()
	defer func() {
		r.metrics.MongoLatency.WithLabelValues("notifications", "find").Observe(time.Since(start).Seconds())
		r.metrics.MongoOperations.WithLabelValues("notifications", "find").Inc()
	}()
	query := bson.M{"user_id": userID}
	total, err := r.collection.CountDocuments(ctx, query)
	if err != nil {
		return nil, 0, err
	}
	if page < 1 {
		page = 1
	}
	if pageSize < 1 {
		pageSize = 20
	}
	opts := options.Find().SetSkip(int64((page - 1) * pageSize)).SetLimit(int64(pageSize)).SetSort(bson.M{"created_at": -1})
	cursor, err := r.collection.Find(ctx, query, opts)
	if err != nil {
		return nil, 0, err
	}
	defer cursor.Close(ctx)
	var notifs []*entity.Notification
	if err := cursor.All(ctx, &notifs); err != nil {
		return nil, 0, err
	}
	return notifs, total, nil
}

func (r *NotificationRepository) MarkAsRead(ctx context.Context, id string) error {
	_, err := r.collection.UpdateOne(ctx, bson.M{"_id": id}, bson.M{
		"$set": bson.M{"read": true, "read_at": time.Now()},
	})
	return err
}

func (r *NotificationRepository) MarkAllAsRead(ctx context.Context, userID string) error {
	_, err := r.collection.UpdateMany(ctx, bson.M{"user_id": userID, "read": false}, bson.M{
		"$set": bson.M{"read": true, "read_at": time.Now()},
	})
	return err
}

func (r *NotificationRepository) CountUnread(ctx context.Context, userID string) (int64, error) {
	return r.collection.CountDocuments(ctx, bson.M{"user_id": userID, "read": false})
}

var _ port.NotificationRepository = (*NotificationRepository)(nil)

File: apps/server/src/adapter/driven/mongodb/notification-repository.ts
import { ObjectId } from "mongodb";
import type { Collection } from "mongodb";
import type { Notification } from "../../../domain/entity/index.js";
import type { NotificationRepository } from "../../../domain/port/index.js";
import type { MongoDBClient } from "./client.js";

interface NotificationDoc {
  _id: ObjectId;
  user_id: string;
  type: string;
  title: string;
  message: string;
  resource_id?: string;
  resource_type?: string;
  read: boolean;
  read_at?: Date;
  created_at: Date;
}

function toDoc(n: Notification): NotificationDoc {
  return {
    _id: new ObjectId(n.id),
    user_id: n.userId,
    type: n.type,
    title: n.title,
    message: n.message,
    resource_id: n.resourceId,
    resource_type: n.resourceType,
    read: n.read,
    read_at: n.readAt,
    created_at: n.createdAt,
  };
}

function fromDoc(d: NotificationDoc): Notification {
  return {
    id: d._id.toHexString(),
    userId: d.user_id,
    type: d.type as Notification["type"],
    title: d.title,
    message: d.message,
    resourceId: d.resource_id,
    resourceType: d.resource_type,
    read: d.read,
    readAt: d.read_at,
    createdAt: d.created_at,
  };
}

export class MongoNotificationRepository implements NotificationRepository {
  private readonly col: Collection<NotificationDoc>;

  constructor(client: MongoDBClient) {
    this.col = client.collection<NotificationDoc>("notifications");
  }

  async findById(id: string): Promise<Notification | null> {
    const doc = await this.col.findOne({ _id: new ObjectId(id) });
    return doc ? fromDoc(doc) : null;
  }

  async findByUserId(userId: string): Promise<Notification[]> {
    const docs = await this.col
      .find({ user_id: userId })
      .sort({ created_at: -1 })
      .toArray();
    return docs.map(fromDoc);
  }

  async findUnreadByUserId(userId: string): Promise<Notification[]> {
    const docs = await this.col
      .find({ user_id: userId, read: false })
      .sort({ created_at: -1 })
      .toArray();
    return docs.map(fromDoc);
  }

  async countUnreadByUserId(userId: string): Promise<number> {
    return this.col.countDocuments({ user_id: userId, read: false });
  }

  async create(notification: Notification): Promise<Notification> {
    await this.col.insertOne(toDoc(notification));
    return notification;
  }

  async markAsRead(id: string): Promise<void> {
    await this.col.updateOne(
      { _id: new ObjectId(id) },
      { $set: { read: true, read_at: new Date() } },
    );
  }

  async markAllAsRead(userId: string): Promise<void> {
    await this.col.updateMany(
      { user_id: userId, read: false },
      { $set: { read: true, read_at: new Date() } },
    );
  }

  async delete(id: string): Promise<void> {
    await this.col.deleteOne({ _id: new ObjectId(id) });
  }
}
